landoapi/landings.py

```diff
@@ -0,0 +1,201 @@
+                job,
+                repo,
+                hgrepo,
+                treestatus_subsystem.client,
+                current_app.config["PATCH_BUCKET_NAME"],
+            )
+
+            # Finalize job
+            db.session.commit()
+            logger.info("Finished processing landing job", extra={"id": job.id})
+            self.job_processing = False
+        logger.info("Landing worker exited")
+
+    def exit_gracefully(self, *args):
+        logger.info(f"Landing worker exiting gracefully {args}")
+        while self.job_processing:
+            time.sleep(self.seconds_between_jobs)
+        self.running = False
+
+    def run_job(self, job, repo, hgrepo, treestatus, patch_bucket):
+        if not treestatus.is_open(repo.tree):
+            job.transition_status(
+                "defer",
+                message=f"Tree {repo.tree} is closed - retrying later.",
+                commit=True,
+                db=db,
+            )
+            return False
+
+        try:
+            # Set the environment variable the `set_landing_system.py` hg
+            # extension will use to send the push user override.
+            os.environ["AUTOLAND_REQUEST_USER"] = job.requester_email
+
+            with hgrepo:
+                hgrepo.update_repo(repo.pull_path)
+
+                for revision_id, diff_id in job.landing_path:
+                    patch_buf = patches.download(
+                        revision_id,
+                        diff_id,
+                        patch_bucket,
+                        aws_access_key=self.config["AWS_ACCESS_KEY"],
+                        aws_secret_key=self.config["AWS_SECRET_KEY"],
+                    )
+                    hgrepo.apply_patch(patch_buf)
+
+                commit_id = hgrepo.run_hg(["log", "-r", ".", "-T", "{node}"]).decode(
+                    "utf-8"
+                )
+                hgrepo.push(repo.push_path, bookmark=repo.push_bookmark or None)
+
+        except NoDiffStartLine:
+            logger.exception("Patch without a diff start line.")
+            message = (
+                "Lando encountered a malformed patch, please try again. "
+                "If this error persists please file a bug."
+            )
+            job.transition_status("fail", message=message, commit=True, db=db)
+            return True
+        except PatchConflict as exc:
+            message = (
+                "We're sorry, Lando could not rebase your "
+                "commits for you automatically. Please manually "
+                "rebase your commits and try again.\n\n"
+                f"{str(exc)}"
+            )
+            job.transition_status("fail", message=message, commit=True, db=db)
+            return True
+        except TreeClosed:
+            # TODO when fetching jobs to run, we should fetch ones that are
+            # "in progress" but transiently failed, though we should put them
+            # at the bottom of the stack.
```

Comment:
yes!  that's much closer to ensuring we land things in the order they were requested.

Reply:
I implemented this but forgot to remove the TODO -- jobs with `DEFERRED` status will be fetched and stacked in order. Will remove this TODO!